### PR TITLE
Pokemon sorting added to generation filter

### LINKED ISSUE #14: Pokemon Sorting
Create the filter sorting based on the figma design and make it work with the list of pokemon in the main page
Sorting Filter Button (Icon)

Image

Opens the Sorting Filter popup and when expanded

Image

### Acceptance Criteria

- [x]   Get the icon Image from the style guide and add it to the header of the page.
- [x] When pressing the sorting filter button, a pop-up should be displayed like the picture above.
- [x] The pop-up should have a title, a short description and a table of buttons with the options to sort.
- [x] There will be four options to sort, every option will have its own title: **1. smallest number first**, **2. Highest number first**, **3. A-Z**, **4. Z-A**.
- [x] When any sort option is pressed, the pokemon list should order the pokemon by that sorted option and the pop up should be closed.
- [x] If there's a sorted option selected, it should be marked as red.
- [x] The sorted option cannot be de-selected.
- [x] When the pop-up is opened for the first time, it should open on its collapsed form.
- [x] When the pop-up is expanded, it should show the complete panel.
- [x] When closing the pop-up is like cancelling the action, nothing should change and the pop-up should dissapear.
- [x] Unit tests

### PR DESCRIPTION
**Pull request by adCBueno**
Files added

### CODE REVIEW

File: Pokedex_ios/Assets.xcassets/IconButtons/sortIcon.imageset/Contents.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "sortIcon.png",
```
Comment (jmacboy) [inline Pokedex_ios/Assets.xcassets/IconButtons/sortIcon.imageset/Contents.json]: we're using SVG icons for this things, please change accordingly

File: Pokedex_ios/View/Controller/PokemonList/PokemonListViewController.swift [hunk 2]
```diff
@@ -42,7 +42,11 @@ class PokemonListViewController: UIViewController {
         let generationFilterButton = UIBarButtonItem(image: generationFilterImage, style: .plain,
                                                    target: self, action: #selector(showGenerationFilterPopup))
 
-        navigationItem.rightBarButtonItems = [advancedFilterButton, generationFilterButton]
+        let sortFilterImage = UIImage(named: "sortIcon")?.withRenderingMode(.alwaysOriginal)
```
Comment (jmacboy) [inline Pokedex_ios/View/Controller/PokemonList/PokemonListViewController.swift:45]: Please use the "sortIcon" name as a constant

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            button.isSelected = !button.isSelected
+            if button.isSelected {
+               button.tintColor = UIColor.white
+               button.backgroundColor = UIColor(named: "apply-button")
+            } else {
+               button.backgroundColor = UIColor.systemGray6
+            }
+        }
+            sortPokemonListViewModel.sortWithSmallestNumberFirst()
+        delayDismiss()
```
Comment (jmacboy) [inline Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift:78]: please indent this file correctly

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+    @IBAction func highestNumberFirst(_ sender: Any) {
+        if let button: UIButton = sender as? UIButton {
+             button.isSelected = !button.isSelected
+            if button.isSelected {
+               button.tintColor = UIColor.white
+               button.backgroundColor = UIColor(named: "apply-button")
```
Comment (jmacboy) [inline Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift]: Please use the constant for this color name

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+          sortPokemonListViewModel.sortWithHighestNumberFirst()
+          delayDismiss()
+        }
+
+    @IBAction func aToZ(_ sender: Any) {
+       if let button: UIButton = sender as? UIButton {
+            button.isSelected = !button.isSelected
+           if button.isSelected {
+              button.tintColor = UIColor.white
+              button.backgroundColor = UIColor(named: "apply-button")
```
Comment (jmacboy) [inline Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift]: also here for the constant, it is already created

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+          sortPokemonListViewModel.sortWithAtoZ()
+          delayDismiss()
+      }
+
+     @IBAction func zToA(_ sender: Any) {
+      if let button: UIButton = sender as? UIButton {
+           button.isSelected = !button.isSelected
+                if button.isSelected {
+                   button.tintColor = UIColor.white
+                   button.backgroundColor = UIColor(named: "apply-button")
```
Comment (jmacboy) [inline Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift]: also here for the constant, it is already created

File: Pokedex_ios/ViewModel/SortPokemonListViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+    var pokemons: [PokemonRaw]
+    weak var sortButtonActionDelegateDelegate: SortButtonActionDelegateDelegate?
+
+    init(pokemons: [PokemonRaw]) {
+        self.pokemons = pokemons
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Comment (jmacboy) [inline Pokedex_ios/ViewModel/SortPokemonListViewModel.swift]: Remove this init if its not needed or its not implemented

File: Pokedex_ios/ViewModel/SortPokemonListViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    var closePopup: (() -> Void)?
+        @objc func applyFilters() {
+            closePopup?()
+        }
+    // MARK: Sorting Logic
+    func sortWithSmallestNumberFirst() {
+        pokemons = pokemons.sorted(by: { pokemanRow1, pokemanRow2 in
```
Comment (jmacboy) [inline Pokedex_ios/ViewModel/SortPokemonListViewModel.swift]: typo in pokemon

File: Pokedex_ios/ViewModel/SortPokemonListViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            return pokemanRow1.id < pokemanRow2.id
+        })
+        sortButtonActionDelegateDelegate?.sortButtonAction(pokemons: pokemons)
+    }
+    func sortWithHighestNumberFirst() {
+        pokemons = pokemons.sorted(by: { pokemanRow1, pokemanRow2 in
```
Comment (jmacboy) [inline Pokedex_ios/ViewModel/SortPokemonListViewModel.swift]: typo in pokemon

File: Pokedex_ios/ViewModel/SortPokemonListViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+            return pokemanRow1.id > pokemanRow2.id
+        })
+        sortButtonActionDelegateDelegate?.sortButtonAction(pokemons: pokemons)
+    }
+    func sortWithAtoZ() {
+        pokemons = pokemons.sorted(by: { pokemanRow1, pokemanRow2 in
```
Comment (jmacboy) [inline Pokedex_ios/ViewModel/SortPokemonListViewModel.swift]: typo in pokemon

File: Pokedex_ios/ViewModel/SortPokemonListViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+            return pokemanRow1.name < pokemanRow2.name
+        })
+        sortButtonActionDelegateDelegate?.sortButtonAction(pokemons: pokemons)
+    }
+    func sortWithZtoA() {
+        pokemons = pokemons.sorted(by: { pokemanRow1, pokemanRow2 in
```
Comment (jmacboy) [inline Pokedex_ios/ViewModel/SortPokemonListViewModel.swift]: typo in pokemon

**Review by jmacboy — CHANGES_REQUESTED** [review]

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.xib [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+                                <action selector="zToA:" destination="-1" eventType="touchUpInside" id="BRK-TG-L66"/>
+                            </connections>
+                        </button>
+                    </subviews>
+                </stackView>
+            </subviews>
+            <viewLayoutGuide key="safeArea" id="fnl-2z-Ty3"/>
+            <color key="backgroundColor" systemColor="systemBackgroundColor"/>
+            <constraints>
+                <constraint firstItem="EoH-0K-Bjl" firstAttribute="leading" secondItem="fnl-2z-Ty3" secondAttribute="leading" constant="20" id="2zz-bp-duA"/>
```
Comment (Rodrigo-Schar) [inline Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.xib]: check the spacing for your stackView it has to be similar as the other filters, change it to 40

**Comment by Rodrigo-Schar** [comment]
Captura de pantalla (15)

Also check the title style and the description textColor, check it from the other filters it has to be similar

File: Pokedex_ios/Assets.xcassets/IconButtons/sortIcon.imageset/Contents.json (hunk 1 above)
Comment (adCBueno) [reply Pokedex_ios/Assets.xcassets/IconButtons/sortIcon.imageset/Contents.json]: Corrected :)

File: Pokedex_ios/View/Controller/PokemonList/PokemonListViewController.swift (hunk 2 above)
Comment (adCBueno) [reply Pokedex_ios/View/Controller/PokemonList/PokemonListViewController.swift:45]: I already checked, it's been used just once :)

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift (hunk 3 above)
Comment (adCBueno) [reply Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift:78]: Corrected :)

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift (hunk 4 above)
Comment (adCBueno) [reply Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift]: Corrected :) tell if it's okay so I canchange it again

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift (hunk 5 above)
Comment (adCBueno) [reply Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift]: Corrected :)

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift (hunk 6 above)
Comment (adCBueno) [reply Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift]: Corrected :)

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift (hunk 4 above)
Comment (adCBueno) [reply Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift]: Corrected :)

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift (hunk 5 above)
Comment (adCBueno) [reply Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift]: Corrected :)

File: Pokedex_ios/ViewModel/SortPokemonListViewModel.swift (hunk 7 above)
Comment (adCBueno) [reply Pokedex_ios/ViewModel/SortPokemonListViewModel.swift]: Checked and indeed it was not necessary, removed already

File: Pokedex_ios/View/Controller/PokemonList/PokemonListViewController.swift (hunk 2 above)
Comment (adCBueno) [reply Pokedex_ios/View/Controller/PokemonList/PokemonListViewController.swift:45]: > Captura de pantalla (15)
> 
> Also check the title style and the description textColor, check it from the other filters it has to be similar

Nice observation, done

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+       }
+
+       func initViewModel() {
+           sortPokemonListViewModel.closePopup = {[weak self] in
+               self?.animateDismissView()
+           }
+       }
+
+       required init?(coder: NSCoder) {
+           fatalError("init(coder:) has not been implemented")
```
Comment (jmacboy) [inline Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift:43]: remove this, if its empty add this:
        super.init(coder: aDecoder)

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+             } else {
+               button.backgroundColor = UIColor.unAppliedButtonColor()
+             }
+         }
+         sortPokemonListViewModel.sortWithZtoA()
+         delayDismiss()
+       }
+}
+
+extension UIColor {
```
Comment (jmacboy) [inline Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift]: as this is an UIColor extension, does not belong here, if you´re going to add this, add it to a different class in a folder called extensions or utils.
If not, you´re breaking the Single Responsibility Principle

Comment (adCBueno) [reply Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift]: True hehe, corrected, thanks :)

File: Pokedex_ios/Utilities/Extentions/UIColor.swift [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+//  Pokedex_ios
+//
+//  Created by adri on 7/19/22.
+//
+import UIKit
+import Foundation
+
+extension UIColor {
+    class func applyButtonColor() -> UIColor {
+        return UIColor(named: "apply-button") ?? UIColor.systemGray6
```
Comment (jmacboy) [inline Pokedex_ios/Utilities/Extentions/UIColor.swift]: also use ConstantVariables.applyButtonColor instead of the string

**Review by Rodrigo-Schar — APPROVED** [review]

File: Pokedex_ios/ViewModel/SortPokemonListViewModel.swift (hunk 8 above)
Comment (adCBueno) [reply Pokedex_ios/ViewModel/SortPokemonListViewModel.swift]: Can you explain a litlle bit more this one please? I'm not sure what I have to correct

File: Pokedex_ios/Utilities/Extentions/UIColor.swift (hunk 15 above)
Comment (adCBueno) [reply Pokedex_ios/Utilities/Extentions/UIColor.swift]: Corrected :)

File: Pokedex_ios/ViewModel/SortPokemonListViewModel.swift (hunk 8 above)
Comment (adCBueno) [reply Pokedex_ios/ViewModel/SortPokemonListViewModel.swift]: Corrected :)

File: Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift (hunk 13 above)
Comment (adCBueno) [reply Pokedex_ios/View/Controller/SortFilter/SortFilterViewController.swift:43]: Removed :)

**Review by jmacboy — APPROVED** [review]

**Pull request merged**